### PR TITLE
[Open_Source_LLM] Add Dynamic List Functionalities

### PR DESCRIPTION
**Pull request by prakharg-msft**
# Description
Support dynamic list functionality to pick endpoint or custom connection for open source llm tool. 
Shoutout to @gjwoods for base implementation for this in https://github.com/microsoft/promptflow/pull/843 
image

Please add an informative description that covers that changes made by the pull request and link all relevant issues.

# All Promptflow Contribution checklist:
- [ ] **The pull request does not introduce [breaking changes].**
- [ ] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [ ] **I have read the [contribution guidelines](../CONTRIBUTING.md).**
- [ ] **Create an issue and link to the pull request to get dedicated review from promptflow team. Learn more: [suggested workflow](../CONTRIBUTING.md#suggested-workflow).**

## General Guidelines and Best Practices
- [x] Title of the pull request is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### Testing Guidelines
- [x] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: src/promptflow-tools/promptflow/tools/open_source_llm.py [hunk 1]
```diff
@@ -51,6 +52,329 @@ def wrapper(*args, **kwargs):
...
+    def get_endpoint_from_azure_custom_connection(self,
+                            subscription_id,
+                            resource_group_name,
+                            workspace_name,
+                            connection_name) -> Tuple[str, str, str]:
+        from promptflow.azure import PFClient
+        from azure.identity import DefaultAzureCredential
+        credential = DefaultAzureCredential(exclude_interactive_browser_credential=False)
+
+        azure_pf_client = PFClient(
```
Comment (chjinche) [inline src/promptflow-tools/promptflow/tools/open_source_llm.py:174]: @brynn-code to help review the way to get connection.

File: src/promptflow/promptflow/_core/connection_manager.py [hunk 2]
```diff
@@ -55,7 +55,7 @@ def _build_connections(cls, _dict: Dict[str, dict]):
                 secret_keys = connection_dict.get("secret_keys", [])
                 secrets = {k: v for k, v in value.items() if k in secret_keys}
                 configs = {k: v for k, v in value.items() if k not in secrets}
-                connection_value = connection_class(configs=configs, secrets=secrets)
+                connection_value = connection_class(name=key, configs=configs, secrets=secrets)
```
Comment (chjinche) [inline src/promptflow/promptflow/_core/connection_manager.py:58]: @brynn-code and @thy09 to help comment on this `core` change.

File: src/promptflow-tools/tests/test_open_source_llm.py [hunk 3]
```diff
@@ -209,67 +225,53 @@ def test_open_source_llm_llama_parse_chat_with_comp(self):
...
-                response = os_llm.call(self.chat_prompt, API.CHAT, deployment_name=deployment_name)
+                response = self.stateless_os_llm.call(
+                    self.chat_prompt,
+                    API.CHAT,
+                    connection=f"onlineEndpoint/{endpoint_name}",
+                    deployment_name=deployment_name)
                 assert len(response) > 25
 
-    @pytest.mark.skip_if_no_api_key("open_source_llm_ws_service_connection")
+    @pytest.mark.skip_if_no_key("open_source_llm_ws_service_connection")
```
Comment (chjinche) [inline src/promptflow-tools/tests/test_open_source_llm.py:249]: please kindly use `skip_if_no_api_key` see https://github.com/microsoft/promptflow/pull/935
maybe you should merge/rebase main branch?

File: src/promptflow-tools/requirements.txt [hunk 4]
```diff
@@ -1,2 +1,3 @@
 google-search-results==2.4.1
 promptflow
+azure-ai-ml
```
Comment (chjinche) [inline src/promptflow-tools/requirements.txt:3]: pf-tools should have not dependency on azure by default. please remove change in requirements.txt.

File: src/promptflow/promptflow/_core/connection_manager.py (hunk 2 above)
Comment (prakharg-msft) [reply src/promptflow/promptflow/_core/connection_manager.py:58]: Sure, without this change connection manager does not return the correct name of the connection. See the highlighted section in the screenshot below.
image

File: src/promptflow-tools/requirements.txt (hunk 4 above)
Comment (prakharg-msft) [reply src/promptflow-tools/requirements.txt]: @chjinche , we need this to refer to MLClient and pull the online endpoints to show in the dropdown. If I remove it from here then the test will start to fail.

image

File: src/promptflow-tools/promptflow/tools/open_source_llm.py (hunk 1 above)
Comment (prakharg-msft) [reply src/promptflow-tools/promptflow/tools/open_source_llm.py:172]: @brynn-code , my tests are running fine locally using connections.json. However when they run on PR they fail with this error. Am I doing something wrong in how locall connections are being pulled? I think CI test setup does not recognize these as locall connections.
image

File: src/promptflow-tools/requirements.txt (hunk 4 above)
Comment (chjinche) [reply src/promptflow-tools/requirements.txt:3]: The common practice is to add 'azure' related dependency as an `extra_requires`. You may see this as an reference https://github.com/microsoft/promptflow/blob/ef22b0f8727623a6190575c1770270fa04c891f7/examples/tools/tool-package-quickstart/setup.py#L16

Comment (chjinche) [reply src/promptflow-tools/requirements.txt:3]: as for tool test issue, you may add the dependency you need to tool_tests, like this https://github.com/microsoft/promptflow/blob/ef22b0f8727623a6190575c1770270fa04c891f7/.github/workflows/tools_tests.yml#L27

File: src/promptflow-tools/tests/test_open_source_llm.py [hunk 5]
```diff
@@ -1,23 +1,33 @@
 import copy
 import os
 import pytest
+from typing import List, Dict
```
Comment (chjinche) [inline src/promptflow-tools/tests/test_open_source_llm.py:4]: @prakharg-msft could you please take a look at the failed tests in `tool_tests` CI? All oss llm tool tests failure and look similar "xx connection not found".
image

Comment (prakharg-msft) [reply src/promptflow-tools/tests/test_open_source_llm.py:4]: Yes, I think this is related to how connection_Manager works in CI. I am seeking help from @brynn-code for this - https://github.com/microsoft/promptflow/pull/976#discussion_r1379303384

File: src/promptflow-tools/requirements.txt (hunk 4 above)
Comment (prakharg-msft) [reply src/promptflow-tools/requirements.txt]: > as for tool test issue, you may add the dependency you need to tool_tests, like this
> 
> https://github.com/microsoft/promptflow/blob/ef22b0f8727623a6190575c1770270fa04c891f7/.github/workflows/tools_tests.yml#L27

Does this change needs to go out in a separate PR? Seems including this change in this PR has no effect on how the CI is setup.

File: src/promptflow-tools/promptflow/tools/open_source_llm.py (hunk 1 above)
Comment (prakharg-msft) [reply src/promptflow-tools/promptflow/tools/open_source_llm.py:172]: Will promptflow.azure be available in default runtime? Is this the right way to get cloud connections?

File: src/promptflow/promptflow/_core/connection_manager.py (hunk 2 above)
Comment (prakharg-msft) [reply src/promptflow/promptflow/_core/connection_manager.py]: This is no longer needed in my tests. I have removed the change.

File: src/promptflow-tools/requirements.txt (hunk 4 above)
Comment (prakharg-msft) [reply src/promptflow-tools/requirements.txt]: @crazygao , help clarify that this was due to merge conflicts on the PR. Resolving those got the package reflected in CI test setup.

File: src/promptflow-tools/tests/test_open_source_llm.py (hunk 5 above)
Comment (prakharg-msft) [reply src/promptflow-tools/tests/test_open_source_llm.py:4]: this is fixed now.

**Review by DaweiCai — DISMISSED** [review]

**Review by chjinche — APPROVED** [review]

File: src/promptflow-tools/promptflow/tools/open_source_llm.py (hunk 1 above)
Comment (chjinche) [reply src/promptflow-tools/promptflow/tools/open_source_llm.py]: Checked `azure-ai-ml` is installed in default runtime.

**Review by brynn-code — APPROVED** [review]

File: src/promptflow-tools/promptflow/tools/open_source_llm.py (hunk 1 above)
Comment (prakharg-msft) [reply src/promptflow-tools/promptflow/tools/open_source_llm.py]: Confirmed over call with, promptflow azure will be available in defauilt runtime and the usage of PFClient is also correct in the tool.

**Pull request merged**